### PR TITLE
Raise errors for inputs that would cause blank outputs; handle length in CLI for variable length digest algorithms

### PR DESCRIPTION
**Pull request by thompsonmj**
Addresses #14 and #17

### CODE REVIEW

**Comment by thompsonmj** [comment]
@zoeduan I made some significant changes to `test_hasher.py` to have parameterize over all test checksums provided in `checksums` with `pytest.mark.parametrize`. It does this for default length settings as well as both with and without a `length` used for BLAKEs. I think this is a helpful reduction of repetition. 

@egrace479 I made it so that if an algorithm is not available on the system, it continues through the loop, skipping the subtest. There's also a test to make sure that if non-guaranteed algs are unavailable, just those that are available are tested.

**Comment by thompsonmj** [comment]
Note that I tested possible failure modes of the testing approach locally by artificially perturbing available and guaranteed sets, e.g.:
```
def is_algorithm_available(algorithm):
    return algorithm in hashlib.algorithms_guaranteed # Only say that an algorithm is available if it's in the guaranteed set
```
Or by removing or writing an error for one or more entries in `checksums` causes tests to fail in the correct places.

File: tests/test_hasher.py [hunk 1]
```diff
@@ -1,140 +1,111 @@
...
+    temp_file.close()
+    yield temp_file_path
+    os.remove(temp_file_path)
+
+@pytest.mark.parametrize("algorithm", checksums.keys())
+def test_all_algorithms_with_defaults(temp_file, algorithm):
+    if not is_algorithm_available(algorithm):
+        if is_algorithm_guaranteed(algorithm):
+            pytest.fail(f"{algorithm} is guaranteed but not available.")
+        pytest.skip(f"{algorithm} not available on system and not guaranteed.")
```
Comment (thompsonmj) [inline tests/test_hasher.py:54]: This skips the test for an algorithm parameterized with `parametrize` in a case like what was discovered with `ripemd160` on the GitHub Actions environment due to some algorithms being made unavailable by OpenSSL by default (as mentioned [here](https://github.com/Imageomics/sum-buddy/pull/15#issue-2430725466)).

File: src/sumbuddy/__main__.py [hunk 2]
```diff
@@ -69,7 +75,9 @@ def main():
         if overwrite.lower() != 'y':
             sys.exit("Exited without executing")
         
-    get_checksums(args.input_dir, args.output_file, args.ignore_file, args.include_hidden, args.algorithm)
-
+    try:
+        get_checksums(args.input_dir, args.output_file, args.ignore_file, args.include_hidden, args.algorithm, args.length)
+    except (LengthUsedForFixedLengthHashError) as e:
+        sys.exit(str(e))
```
Comment (egrace479) [inline src/sumbuddy/__main__.py:81]:
```suggestion
        sys.exit(str(e))


```
Just readability

File: src/sumbuddy/exceptions.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+class NoFilesAfterFilteringError(Exception):
+    def __init__(self, input_directory, ignore_file):
+        message = f"The directory {input_directory} contains files, but all are filtered out. \nCheck patterns in your {ignore_file} file and/or hidden files settings."
+        super().__init__(message)
+
+class LengthUsedForFixedLengthHashError(Exception):
+    def __init__(self, algorithm):
+        message = f"Length paremeter is not applicable for fixed-length algorithm '{algorithm}'."
+        super().__init__(message)
```
Comment (egrace479) [inline src/sumbuddy/exceptions.py:14]:
```suggestion
        super().__init__(message)

```

File: src/sumbuddy/exceptions.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
+class EmptyInputDirectoryError(Exception):
+    def __init__(selfe, input_directory):
```
Comment (egrace479) [inline src/sumbuddy/exceptions.py:2]:
```suggestion
    def __init__(self, input_directory):
```

**Review by egrace479 — CHANGES_REQUESTED** [review]

**Review by zoeduan — COMMENTED** [review]
I run the test files below: 
1. test.filter.py - all passed
2. test_getChecksums.py - all passed
3. test_mappper.py - all passed
4. test_hasher.py - please refer to log below.
collected 42 items                                                                                                                                                           

test_hasher.py .................................sssss....                                                                                                              [100%]

======================================================================= 37 passed, 5 skipped in 0.04s

**Review by zoeduan — APPROVED** [review]

File: tests/test_hasher.py [hunk 5]
```diff
@@ -1,140 +1,111 @@
...
+
+    if algorithm in blake_algorithms_with_defaults:
+        length = None
+        checksum = hasher.checksum_file(temp_file, algorithm=algorithm)
+        assert checksum == checksums[algorithm], f"Assertion failed for {algorithm} with length {length}"
+
+@patch('hashlib.algorithms_available', new_callable=lambda: hashlib.algorithms_guaranteed)
+@pytest.mark.parametrize("algorithm", checksums.keys())
+def test_only_guaranteed_algorithms_available(mock_algorithms, temp_file, algorithm):
+    test_all_algorithms_with_defaults(temp_file, algorithm)
```
Comment (egrace479) [inline tests/test_hasher.py:80]:
```suggestion
    '''
    Run algorithm tests as if those that are not guaranteed are unavailable in the system.
    Skips are expected, but each algorithm should only be skipped once unless it is actually unavailable on the system running the testing.
    '''
    test_all_algorithms_with_defaults(temp_file, algorithm)
```

**Review by egrace479 — APPROVED** [review]

**Pull request merged**